ark/phenotyping/preprocess.py

```diff
@@ -10,6 +11,60 @@
 from ark.utils import load_utils
 from ark.utils import misc_utils
 
+from timeit import default_timer
+
+
+def _unpacking_apply_along_axis(func_args):
+    """Maps apply_along_axis to each array subset
+
+    Args:
+        func_args (tuple):
+            The arguments to pass into apply_along_axis
+
+    Returns:
+        numpy.ndarray:
+            The result for apply_along_axis for a specific array subset
+    """
+
+    (func1d, axis, arr, args, kwargs) = func_args
+    return np.apply_along_axis(func1d, axis, arr, *args, **kwargs)
+
+
+def _parallel_apply_along_axis(func1d, axis, arr, *args, **kwargs):
+    """Runs apply_along_axis in parallel for a massive speedup (but at a massive memory cost too)
+
+    Args:
+        func1d (Callable):
+            The function to apply along the given axis
+        axis (int):
+            The axis to run
+
+    Returns:
+        numpy.ndarray:
+            The processed array, equivalent to what np.apply_along_axis returns
+    """
+
+    # If axis is 0 we'll effectively be working with the 1st axis
+    effective_axis = 1 if axis == 0 else axis
+
+    # Swap the 0th and 1st axis if axis specified is 0
+    if effective_axis != axis:
+        arr = arr.swapaxes(axis, effective_axis)
+
+    # Map the array subsets
+    chunks = [(func1d, effective_axis, sub_arr, args, kwargs)
```

Comment:
cpu_count is a good default, but we should add in an optional parameter to the main function which allows the user to specify a smaller number of parallel threads in case they run into memory issues. 

Correct me if I'm wrong, but the actual amount of memory being used here is not that substantial, right? We're talking about at most 8 images being run simultaneously? This seems fairly insignificant when compared to loading an xarray containing an entire cohort of images

Reply:
I think the `Pool` here is actually a process pool, which is slightly different than a threadpool.  Instead of everything running on a thread, everything is a different system process.  Basically, each process is slower, but you can get more than the cpu core limit I believe.

As for the memory, I think it's being duplicated (not referenced) when the chunks are made, and the float64s don't help that. Loading an entire cohort of images makes this seem not bad, but most of the time, those images are uint8 or uint16 at worse.  When you load in float64 tiffs (mibitiffs default to this 🙄), it's a different story -> ~(8*num_channels) MB per image.  If you have 20 channels, 7 images already occupy >1GB of ram.

All that to say, I'm pretty sure that the bluring defaults to outputting float64 arrays.  Rescaling and converting these into uint16s or even uint8s would probably help (either before or after bluring).